### PR TITLE
[PoW Merge] Move txn sharing assignments serialization code

### PR DESCRIPTION
**Pull request by ansnunez**
## Description

Another PR in the PoW merging work.

This is a similar effort to #464 where sharding info serialization was moved to DSBlockMessage.h.
This time, serialization of txn sharing assignments has been moved to that same file.

## Review Suggestion

What I basically did was split `AppendSharingSetupToShardingStructure()` into two functions:
- `DirectoryService::ComputeTxnSharingAssignments()` handles the computation part
- `TxnSharingAssignments::Serialize()` handles the serialization part

The assignment algorithm remains the same, except now we have to store the result of the computation part so that we can pass it to the serialization part.  Hence, 3 new member variables `m_DSReceivers`, `m_shardReceivers`, and `m_shardSenders`.

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test
- [ ] ~~small-scale cloud test~~

### CODE REVIEW

File: src/libDirectoryService/ShardingPreProcessing.cpp [hunk 1]
```diff
@@ -88,131 +88,82 @@ void DirectoryService::ComputeSharding()
...
     for (unsigned int i = 0; i < m_shards.size(); i++)
     {
         const map<PubKey, Peer>& shard = m_shards.at(i);
 
-        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                  "Shard " << i << " forwarders:");
-
         // PART 2
+
+        m_shardReceivers.emplace_back(vector<Peer>());
```
Comment (ckyang) [inline src/libDirectoryService/ShardingPreProcessing.cpp]: The reason we recommend emplace_back is to reduce the object construction overhead. Thus it's better to just write m_shardReceivers.emplace_back() here.

File: src/libDirectoryService/ShardingPreProcessing.cpp [hunk 2]
```diff
@@ -88,131 +88,82 @@ void DirectoryService::ComputeSharding()
...
+            m_shardReceivers.back().emplace_back(node_peer->second);
             node_peer++;
         }
 
-        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                  "Shard " << i << " senders:");
-
         // PART 3
+
+        m_shardSenders.emplace_back(vector<Peer>());
```
Comment (ckyang) [inline src/libDirectoryService/ShardingPreProcessing.cpp]: The reason we recommend emplace_back is to reduce the object construction overhead. Thus it's better to just write m_shardReceivers.emplace_back() here.

File: src/libDirectoryService/ShardingPreProcessing.cpp [hunk 3]
```diff
@@ -88,131 +88,82 @@ void DirectoryService::ComputeSharding()
...
             nodes_recv_hi = shard.size() - 1;
         }
 
         unsigned int num_nodes = nodes_recv_hi - nodes_recv_lo + 1;
 
-        Serializable::SetNumber<uint32_t>(sharding_structure, curr_offset,
-                                          num_nodes, sizeof(uint32_t));
-        curr_offset += sizeof(uint32_t);
-
         map<PubKey, Peer>::const_iterator node_peer = shard.begin();
```
Comment (ckyang) [inline src/libDirectoryService/ShardingPreProcessing.cpp]: Maybe we can use "auto node_peer" to make code simpler?

File: src/libDirectoryService/ShardingPreProcessing.cpp (hunk 1 above)
Comment (ansnunez) [reply src/libDirectoryService/ShardingPreProcessing.cpp]: Done

File: src/libDirectoryService/ShardingPreProcessing.cpp (hunk 2 above)
Comment (ansnunez) [reply src/libDirectoryService/ShardingPreProcessing.cpp]: Done

File: src/libDirectoryService/ShardingPreProcessing.cpp (hunk 3 above)
Comment (ansnunez) [reply src/libDirectoryService/ShardingPreProcessing.cpp]: Done

**Pull request merged**